includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -1952,73 +1969,96 @@ static function( $matches ) {
-	 * @return array Validation errors.
+	 * @return array {
+	 *     Processed CSS list.
+	 *
+	 *     @type array $validation_errors
+	 *     @type array $viewport_rules
+	 *     @type array $imported_font_urls
+	 * }
 	 */
 	private function process_css_list( CSSList $css_list, $options ) {
-		$results = [];
+		$validation_errors  = [];
+		$viewport_rules     = [];
+		$imported_font_urls = [];
 
 		foreach ( $css_list->getContents() as $css_item ) {
 			$sanitized = false;
 			if ( $css_item instanceof DeclarationBlock && empty( $options['validate_keyframes'] ) ) {
-				$results = array_merge(
-					$results,
+				$validation_errors = array_merge(
+					$validation_errors,
 					$this->process_css_declaration_block( $css_item, $css_list, $options )
 				);
 			} elseif ( $css_item instanceof AtRuleBlockList ) {
 				if ( ! in_array( $css_item->atRuleName(), $options['allowed_at_rules'], true ) ) {
-					$error     = [
+					$error               = [
 						'code'      => self::CSS_SYNTAX_INVALID_AT_RULE,
 						'at_rule'   => $css_item->atRuleName(),
 						'type'      => AMP_Validation_Error_Taxonomy::CSS_ERROR_TYPE,
 						'spec_name' => $options['spec_name'],
 					];
-					$sanitized = $this->should_sanitize_validation_error( $error );
-					$results[] = compact( 'error', 'sanitized' );
+					$sanitized           = $this->should_sanitize_validation_error( $error );
+					$validation_errors[] = compact( 'error', 'sanitized' );
 				}
 				if ( ! $sanitized ) {
-					$results = array_merge(
-						$results,
-						$this->process_css_list( $css_item, $options )
+					$at_rule_processed_list = $this->process_css_list( $css_item, $options );
+					if ( ! empty( $at_rule_processed_list['viewport_rules'] ) ) {
+						$viewport_rules[] = $at_rule_processed_list['viewport_rules'];
+					}
+
+					$validation_errors = array_merge(
+						$validation_errors,
+						$at_rule_processed_list['validation_errors']
 					);
 				}
 			} elseif ( $css_item instanceof Import ) {
-				$results = array_merge(
-					$results,
-					$this->splice_imported_stylesheet( $css_item, $css_list, $options )
-				);
+				$imported_stylesheet = $this->splice_imported_stylesheet( $css_item, $css_list, $options );
+				$imported_font_urls  = array_merge( $imported_font_urls, $imported_stylesheet['imported_font_urls'] );
+				$validation_errors   = array_merge( $validation_errors, $imported_stylesheet['validation_errors'] );
 			} elseif ( $css_item instanceof AtRuleSet ) {
-				if ( ! in_array( $css_item->atRuleName(), $options['allowed_at_rules'], true ) ) {
-					$error     = [
+				if ( in_array( $css_item->atRuleName(), $this->allowed_viewport_rules, true ) ) {
+					$output_format = new OutputFormat();
+					foreach ( $css_item->getRules() as $rule ) {
+						$rule_value = $rule->getValue();
+						if ( $rule_value instanceof Value ) {
+							$rule_value = $rule_value->render( $output_format );
+						}
+
+						$viewport_rules[ $rule->getRule() ] = $rule_value;
```

Comment:
What else would it be? If it is not a `Value`, should it be added to `$viewport_rules`?

Reply:
In several unit tests, `$rule_value` is a string, like `device-width`

The logic above is [taken from](https://github.com/sabberworm/PHP-CSS-Parser/blob/a80a97d3fdb2164d72cbf0fb3cfc40b1d630a877/lib/Sabberworm/CSS/Rule/Rule.php#L200-L204

In [AMP_Style_Sanitizer_Test](https://github.com/ampproject/amp-wp/blob/88a136b484ff8a9c0d30986ce95a739a3e70fafc/tests/php/test-amp-style-sanitizer.php), there were 3 times when `$rule_value instanceof Value` was `true`, and 8 times where it was `false`.

Reply:
Yes, I think it should be added to `$vewport_rules` if it's not a `Value`.

Reply:
Thanks. Good explanation. Makes sense.